### DISCUSSION TITLE
Graclus weight

### DISCUSSION THREAD

**Discussion by mapapin** [pyg-team/pytorch_geometric#4879] [Q&A]
Hi,
I'm trying to do a cluster by edge attribute using graclus function but it returns 
```
torch.cumsum(deg, 0, out=rowptr[1:])

    return torch.ops.torch_cluster.graclus(rowptr, col, weight)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
RuntimeError: optional_weight.value().dim() == 1INTERNAL ASSERT FAILED at "csrc/cpu/graclus_cpu.cpp":12, please report a bug to PyTorch. Input mismatch
```

My edge attribute have a shape of [num_edges, 1] and edge_index [2, num_edges] and i call graclus like that :
`cluster = graclus(data.edge_index, weight=data.edge_attr, num_nodes=data.x.size(0))`

Any suggestions are welcome :)

And then, not related to the error, what am I supposed to have in the pos attribute of a Data object?

Thanks !

**Comment by wsad1** [comment]
Is the shape of `data.edge_attr` [num_edges,1]. `graclus` only supports a 1-d `weight` argument.

**Reply by mapapin** [reply]
Thanks for your answer ! Indeed, it's in the documentation... I must have confused with another function documentation...
So, if I have several features in `edge_attr`, do I combine them to make a vector of dimension `num_edges` ?

**Reply by wsad1** [reply]
Not an expert on this. But I guess you could try that. Are you like planning on taking an average of the `edge_attr` or something?

**Reply by mapapin** [reply]
Yes, I only have one feature at the moment, so I can easily put the vector in one dimension but it is likely that later I will have several.

I did some tests and it seems that graclus runs infinitely when I give it `weight=data.edge_attr.view(-1)` with a GPU. Even if my dataset is small. Is it a common problem ? Should i open an issue ?

Thanks for your answers. :)

**Reply by rusty1s** [reply]
METIS does not have a GPU implementation. There are also some tricks involved to convert `edge_weight` internally since METIS can only handle discrete edge weights. If you have a reproducible example, I am happy to look into this.

**Reply by mapapin** [reply]
Oh ok, that's why my model gets stuck in graclus ? My edge_attr are equivalent to euclidean distance between nodes.

I'm a newbie, maybe this isn't the way to do it but I wanted to make graclus create a cluster based on the closest nodes in space for pooling purpose. 
I have my Data object : `DataBatch(x=[30622, 1], edge_index=[2, 258078], edge_attr=[258078, 1], tag=[64], ids=[30622], num_nodes=30622, batch=[30622], ptr=[65]` and i send data to graclus like that : `cluster = graclus(data.edge_index, weight=data.edge_attr.view(-1), num_nodes=data.x.size(0))` with data previously sent to my convolution.

All my Data objects are sent to GPU before i start training.

I hope that's enough to see what i'm trying to do...

Thanks :)

**Reply by rusty1s** [reply]
This shouldn't be a problem. We auto-convert to CPU inside METIS, see [here](https://github.com/rusty1s/pytorch_sparse/blob/master/torch_sparse/metis.py#L33-L40).

**Reply by mapapin** [reply]
It's weird, when I send my data to the GPU it doesn't work, whereas when I send it to the CPU and it works.

Is it possible that the conversion takes  a very long time even if i send a very small amount of data (10 samples) ?

NB: If I remove graclus, my code works with GPU, it's seems really stuck in the function.

**Reply by rusty1s** [reply]
Can you do some debugging on your end, e.g., where exactly it gets stuck in the `metis.py` function in `torch-sparse`?
